core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Migration\Version411;
+
+use Contao\CoreBundle\Migration\AbstractMigration;
+use Contao\CoreBundle\Migration\MigrationResult;
+use Doctrine\DBAL\Connection;
+
+/**
+ * @internal
+ */
+class TwoFactorBackupCodesMigration extends AbstractMigration
+{
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    public function shouldRun(): bool
+    {
+        $schemaManager = $this->connection->getSchemaManager();
+
+        if (!$schemaManager->tablesExist(['tl_user', 'tl_member'])) {
+            return false;
+        }
+
+        $userColumns = $schemaManager->listTableColumns('tl_user');
+        $memberColumns = $schemaManager->listTableColumns('tl_member');
+
+        if (!isset($userColumns['backupcodes'], $memberColumns['backupcodes'])) {
+            return false;
+        }
+
+        $users = $this->getAffectedRowsForTable('tl_user');
+        $members = $this->getAffectedRowsForTable('tl_member');
+
+        return \count($users) || \count($members);
+    }
+
+    public function run(): MigrationResult
+    {
+        $this->updateBackupCodes('tl_user');
+        $this->updateBackupCodes('tl_member');
+
+        return $this->createResult(true);
+    }
+
+    private function updateBackupCodes(string $table): void
+    {
+        $rows = $this->getAffectedRowsForTable($table);
+
+        foreach ($rows as $row) {
+            $backupCodes = json_decode($row['backupCodes'], true);
```

Comment:
What if `json_decode()` fails? Do we have to handle this case?

Reply:
Should we ignore JSON errors here so that the migration can update as many rows as possible?
```suggestion
            $backupCodes = json_decode($row['backupCodes'], true);

            if (!\is_array($backupCodes)) {
                continue;
            }
```

Reply:
Added in f9d23fe696eee867b88dd5359fe45ce20c3a5b85.